File: add_workers.jl
#!/usr/bin/env julia

# Lets add workers programatically, and see if they distribute to different cores.


# using ArgParse
#
# s = ArgParseSettings()
# @add_arg_table s begin
#     # "--opt1"
#     # help = "an option with an argument"
#     "--p", "-p"
#     help = "number of processes to add"
#     arg_type = Int
#     default = 0
# end
#
# parsed_args = parse_args(ARGS, s)
#
# p = parsed_args["p"]
#
# if p > 0
#     addprocs(p)
# end

nchild = nworkers()
println("Workers allocated ", nchild)
println("Total workers ", workers())

println("Total processes ", nprocs())

@everywhere function print_cpus()
    run(pipeline(`ps -eo pid,psr,pcpu,cmd`, `grep julia`))
end

println("Before map")
print_cpus()

# Trying to set path
@everywhere ENV["PATH"] = ENV["PATH"] * "/n/home07/iczekala/.build/radmc-3d/version_0.38/src"


@everywhere function pcpus(mat)
    println("In map")
    print_cpus()
    run(`which radmc3d`)
    svd(mat)
end

M = [rand(1000,1000) for i=1:nchild]

pmap(pcpus, M)

println("After map.")
print_cpus()

File: add_workers.py
#!/usr/bin/env python

# Lets add workers programatically, and see if they distribute to different cores.

import multiprocessing as mp
import numpy as np
import subprocess

import argparse
parser = argparse.ArgumentParser(prog="add_workers.py", description="Test workers.")
parser.add_argument("-p", "--p", help="Number of workers to add.", type=int)
args = parser.parse_args()

p = args.p

def print_cpus():
    subprocess.call("ps -eo pid,psr,pcpu,cmd | grep python", shell=True)

print("Before map")
print_cpus()


def pcpus(mat):
    print("During map")
    print_cpus()
    return mat**2

pool = mp.Pool(args.p)

M = [np.random.rand(10000,1000) for i in range(p)]

pool.map(pcpus, M)

print("After map.")
print_cpus()
